### PR TITLE
SBML refactoring and bug fixes

### PR DESCRIPTION
**Pull request by ayush9pandey**
To address issues in:-
* #129 
* #122 
* #128 
* #127

### CODE REVIEW

**Comment by zoltuz** [comment]
There's a bug, which came from the previous PR, sorry about that! 
Screenshot 2020-07-26 at 12 44 42
There are two options to solve this:
1, use defaultdict, which creates the missing keys in the dictionary on the fly
2, make a copy of ```self.propensity_dict```, just like here:
https://github.com/BuildACell/BioCRNPyler/blob/dev/biocrnpyler/propensities.py#L401

**Comment by ayush9pandey** [comment]
Thanks, Zoltan. i used the proposed solution # 2 and it did fix some of the errors but I am still seeing the key error 'k' at a few places. I will look at how to fix that. 

#122 should be fixed now, also #127 (once the above error is fixed). 

Next commit will do #129.

**Comment by ayush9pandey** [comment]
Opening it for review. The code changes required are close to done..only need to finish up with bug fixes so that all the tests run.

**Comment by zoltuz** [comment]
Hurray! All test passed. You want to add more for the PR @ayush9pandey  or it is ready for review?

**Comment by ayush9pandey** [comment]
The coverage went down I guess because I added the modifier reference code and that is not being tested. Also, the validate SBML functions are being used for testing but are not tested themselves so probably causing the lowered coverage percentage. I won't bother with this for now but we might wanna add simple tests for these things or move them to more appropriate places in the code. 

I will see if I can fix #135 as well with this PR @zoltuz  before requesting review.

**Comment by ayush9pandey** [comment]
General propensity code was also not being tested earlier, so with this next commit I have added some refactored testing code that does that. The bioscrape annotation testing code is now in a new function which makes more sense IMO. 

This commit will probably break things again but should be easy to fix those.

**Comment by zoltuz** [comment]
Hi @ayush9pandey 
Please a write_to_sbml and validate_sbml to https://github.com/BuildACell/BioCRNPyler/blob/dev/Tests/test_mixtures_and_components_combinatorially.py

File: biocrnpyler/propensities.py [hunk 1]
```diff
@@ -167,47 +167,122 @@ def from_dict(cls, propensity_dict):
         merged.update(propensity_dict['species'])
         return cls(**merged)
 
-    def _create_bioscrape_annotation(self, model, propensity_dict_in_sbml):
+    def _create_annotation(self, model, propensity_dict_in_sbml, **kwargs):
+        '''
```
Comment (zoltuz) [inline biocrnpyler/propensities.py:171]: Usually, triple double-quote is used for docstring: """, but you can leave it like this if you want.

File: biocrnpyler/sbmlutil.py [hunk 2]
```diff
@@ -225,18 +223,28 @@ def _create_reactants(reactant_list, sbml_reaction, model):
         # What to do when there are multiple species with same name?
```
Comment (zoltuz) [inline biocrnpyler/sbmlutil.py:223]: Do we want to remove this comment?

File: biocrnpyler/sbmlutil.py [hunk 3]
```diff
@@ -419,3 +426,94 @@ def getSpeciesByName(model, name, compartment=''):
...
+
+def validate_sbml(sbml_document, enable_unit_check = False, print_results = True):
+    """
+    Validates the generated SBML model by using libSBML SBML validation code
+    """
+    validator = validateSBML(enable_unit_check)
+    validation_result = validator.validate(sbml_document, print_results = print_results)
+    if validation_result > 0:
+        warn('SBML model invalid. Run with print_results = False to hide print statements')
+    return validation_result
```
Comment (zoltuz) [inline biocrnpyler/sbmlutil.py:519]: add a new line at the end. pls

File: biocrnpyler/sbmlutil.py (hunk 2 above)
Comment (ayush9pandey) [reply biocrnpyler/sbmlutil.py:223]: Did we do something about this? Is it printed out as an error/warning some place?

File: biocrnpyler/propensities.py (hunk 1 above)
Comment (ayush9pandey) [reply biocrnpyler/propensities.py:171]: Fixed.

File: biocrnpyler/sbmlutil.py (hunk 3 above)
Comment (ayush9pandey) [reply biocrnpyler/sbmlutil.py:519]: Done.

**Comment by ayush9pandey** [comment]
I am working on updating the notebooks so that they all work with the current changes in dev...these are minor things that are causing errors..we can merge the PR once I have checked all the notebooks are working, generating simulation plots, and the SBML files are valid + import-able by other tools (COPASI).

**Comment by ayush9pandey** [comment]
I think this PR is ready to be merged from my end. @WilliamIX  and @zoltuz  please feel free to suggest changes (particularly look at the couple lines here : https://github.com/ayush9pandey/BioCRNPyler/blob/f178458e8c6820e42fd1c8ff709888530c1d2aeb/biocrnpyler/propensities.py#L211 ). With this change - replacing the first appearance of "k_forward" by "k" fixes the bioscrape annotation problem that we discussed today. It is ofcourse hacky, but I think a lot of things with bioscrape annotations are that way. In case you have a better way to solve this or if you think this will break some use case let me know.

**Comment by ayush9pandey** [comment]
Also, @zoltuz  any idea why the codecov/project check failed? That one always seems to fail for me :(

**Comment by WilliamIX** [comment]
> Also, @zoltuz any idea why the codecov/project check failed? That one always seems to fail for me :(

The issue is that you are losing .35% of testing coverage overall. It looks like this is mostly happening in propensities.py and sbmlutil.py - largely this is due to the new validate_sbml function, which is used for testing, so I think this is ok. If you want to pass though, add some more unit tests to cover other parts of the code.

**Comment by ayush9pandey** [comment]
All green! New tests added for SBML validation and libsbml validation.

File: biocrnpyler/sbmlutil.py (hunk 2 above)
Comment (WilliamIX) [reply biocrnpyler/sbmlutil.py:223]: I think we should definitely have a warning

**Pull request merged**